yt_dlp/YoutubeDL.py

```diff
@@ -2017,7 +2017,7 @@ def can_merge():
             else 'bestvideo*+bestaudio/best' if not compat
             else 'bestvideo+bestaudio/best')
 
-    def build_format_selector(self, format_spec):
+    def build_format_selector(self, format_spec, default_output_format=None):
```

Comment:
Why are you changing this again? I thought this matter had already been settled.

Have you checked every other extractor to make sure this change won't have unintended consequences?

Reply:
It seems to me that this is a far more satisfactory and robust approach than overwriting file extensions. The change need not be merged in if you don't like it, though.

I don't know what other extractors intend when including an extension which is silently discarded. Perhaps the result will be more to their liking?

Reply:
How the formats dicts work is it sorts them and picks the most fitting one, updating the base info dict with the format keys. That is why it seems like the `ext` is getting discarded, bc the format dicts `ext` key will overwrite the one in the base dict. Formats should stand for themselves, otherwise sorting them relative to another does not work consistently.

Please do explain why setting the extension for each format is unsatisfactory. It is how the system was designed to work and it seems more robust to me. Currently all extractors will set the extension on the formats either explicitly or implicitly, for easier support of non homogeneous extension in the formats. I see no benefit in changing it like this instead? Is it for convenience?

There was only [one change](https://github.com/yt-dlp/yt-dlp/pull/5840/files/a500e20ade2ccf8bd227b72bc8dc61c62676e07a#r1056789604) pending to get this PR into a mergeable state before reverting the `YoutubeDL` commit, which was conforming the tests to our test runner. Since you keep introducing other, potentially more drastic changes, we need to revise these (anew?) and question those changes.

We are open to changes that benefit the core, which is why I am asking you again and again on how this change is beneficial over our suggested solutions (May it only be convenience). However please understand that we need to understand these changes to be able to merge them into the code. Why not create an issue first explaining the problem, then after the quirks have been ironed out you can open a new PR with the related core changes. Core PRs tend to take a long(er) time to be merged bc of the potential implications. That way this PR can be merged more quickly.

I hope this answered all your questions.

Reply:
**1st and 2nd paragraph:** It is not so simple (at least) for the ism format (as far as I could figure it out). In that case, two formats - one video and one audio - have to be chosen and "merged" into a container. Therefore, none of the formats alone stands for the result. The final extension can be chosen (within certain bounds) - see `--merge-output-format`. So far, the merging procedure is chosen even before the `info_dict` is considered (to catch potential errors, as the comment tells us). But it is possible to include that information in a second pass, adding an optional parameter to `build_format_selector`. I think it would be nice to set the output format according to what the site thinks it should be. It can still be overridden by the user.

Overwriting the (correct) extensions in the formats is a rather strange approach. It works for `mp4`, apparently because that can contain video as well as audio and because the actual `protocol` is specified separately. Subsequently, this also determines the output format for merge, as desired. Is this more than a coincidence?

The contents of the `info_dict` are `update`d not with the final (merged) format but with the video part of the result. Therefore, some entries for the test are different from the json output, see `-j`. In particular, the value for `ext` does not necessarily reflect the actual file extension. Is this a bug?

**4th paragraph:** I'm not at all surprised that you are reluctant to changes at a central component. Having that said, however, I think the changes are not that difficult to understand this time (as opposed to my first try).

It may indeed be better to cancel this PR, because I'm not really satisfied with the code right now and even less with what you want me to do. Let's wait until, if ever, someone else requests support for this site.

Reply:
Do demonstrate and show your wish in a new issue, not here.

Extension is set for each formats because some services provide multiple formats for a video/audio. YouTube is an example.
Most of the time it's not overwritten since base info_dict don't have `ext` set.

What merged extension should it be for ISM? The code is setting to MP4 all the time. It's not used at all. So the merged format must be MOV while downloaded formats are all MP4 if the website thinks so? Ridiculous.



> The contents of the `info_dict` are `update`d not with the final (merged) format but with the video part of the result. Therefore, some entries for the test are different from the json output, see `-j`. In particular, the value for `ext` does not necessarily reflect the actual file extension. Is this a bug?

Test cases in extractors are meant to test the output from extractor itself. Not a bug at all

Reply:
Do demonstrate and show your wish in a new issue, not here.

Extension is set for each formats because some services provide multiple formats for a video/audio. YouTube is an example.
Most of the time it's not overwritten since base info_dict don't have `ext` set.

What merged extension should it be for ISM? The code is setting to MP4 all the time. It's not used at all. So the merged format must be MOV while downloaded formats are all MP4 if the website thinks so? Ridiculous.



> The contents of the `info_dict` are `update`d not with the final (merged) format but with the video part of the result. Therefore, some entries for the test are different from the json output, see `-j`. In particular, the value for `ext` does not necessarily reflect the actual file extension. Is this a bug?

Test cases in extractors are meant to test the output from extractor itself. Not a bug at all

Reply:
**3rd paragraph:** The merged extension should be `mp4`. The possibility to specify it should *remove* the need to overwrite standard extensions in all the formats. Did you read my comment?

You are right that there aren't many alternatives for the container: `mp4`, `mov`, `flv`, `avi`. Each of them is suitable for the intermediate files, too.

**Last paragraph:** Test cases *do not* test the output of the extractor itself. They test a fairly processed outcome of its output. In fact, they see an intermediate result of the processing, somewhat before the final result, which can be seen in the JSON dump. Did you read my comment? Why do I need to explain the functionality of *your* application to you?

And please don't mess with my code. You can do whatever you like during and after a pull.